### DISCUSSION TITLE
generating booklets

### DISCUSSION THREAD

**Discussion by pr-apes** [speedata/publisher#443] [General]
@pgundlach,

I think this would be a way of using Publisher to generate booklets:

``` xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
  xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Pageformat width="297mm" height="210mm"/>

  <Record element="data">
    <Loop select="if (sd:number-of-pages($myfile) mod 4 != 0) then
        (sd:number-of-pages($myfile) + (4 - (sd:number-of-pages($myfile) mod 4))) div 2
      else sd:number-of-pages($myfile) div 2"
    variable="runpage">
      <Switch>
        <Case test="sd:odd($runpage)">
          <Switch>
            <Case test="((sd:number-of-pages($myfile) + ((4-(sd:number-of-pages($myfile)) mod 4) mod 4) - ($runpage - 1)) &lt;= sd:number-of-pages($myfile)">
              <PlaceObject column="0mm" row="0mm">
                <Image file="{$myfile}" width="{$_pagewidth div 2}" page="{sd:number-of-pages($myfile) + ((4-(sd:number-of-pages($myfile)) mod 4) mod 4) - ($runpage - 1)}"/>
              </PlaceObject>
            </Case>
          </Switch>
          <PlaceObject column="{$_pagewidth div 2}" row="0mm">
            <Image file="{$myfile}" width="{$_pagewidth div 2}" page="{$runpage}"/>
          </PlaceObject>
          <ClearPage/>
          <Switch>
            <Case test="$runpage + 1 &lt;= sd:number-of-pages($myfile)">
              <PlaceObject column="0" row="0mm">
                <Image file="{$myfile}" width="{$_pagewidth div 2}" page="{$runpage+1}"/>
              </PlaceObject>
            </Case>
          </Switch>
          <Switch>
            <Case test="((sd:number-of-pages($myfile) + ((4-(sd:number-of-pages($myfile)) mod 4) mod 4) - $runpage) &lt;= sd:number-of-pages($myfile)">
              <PlaceObject column="{$_pagewidth div 2}" row="0mm">
                <Image file="{$myfile}" width="{$_pagewidth div 2}" page="{sd:number-of-pages($myfile) + ((4-(sd:number-of-pages($myfile)) mod 4) mod 4) - $runpage)}"/>
              </PlaceObject>
            </Case>
          </Switch>
          <ClearPage/>
        </Case>
      </Switch>
    </Loop>
  </Record>
  
</Layout>
```

It would be also useful to be able to remove pages and to add single blank pages after and before any pages.

With a filter, it would be easy to create a table with the right page order, but I don't know how to pass the table to the layout and how to loop the page order.

I have just discovered that `<ForAll>` is my friend here (after converting the table to a data source).

This is just a comment. It might make a good recipe for the documentation.

**Comment by pgundlach** [comment]
That is nice. I think I'll put it into the documentation (I need to add some comments for me first).  One small remark. The speedata Publisher has a crappy, a bit too tolerant XML parser. I have edited the code because `<` in `select="..."` needs to be escaped to `&lt;`.

**Reply by pr-apes** [reply]
It seems that `<` is a valid XPath expression operator (https://www.ibm.com/docs/en/idr/11.3.3?topic=functions-xpath-expression-operators and https://www.w3schools.com/xml/xpath_operators.asp).

**Reply by pgundlach** [reply]
That is true, but < must be escaped in XML. See https://www.w3.org/TR/xml11/#NT-AttValue

**Reply by pr-apes** [reply]
I don't know whether XPath expressions were meant to be part of XML code.

But personally, not having to escape `<` or `>` makes the code more readable for me.

I think that the double quotes in attribute values are like Vegas: what happens there, stays there.

**Reply by pgundlach** [reply]
> I don't know whether XPath expressions were meant to be part of XML code.

In XSLT they are, so this is quite a common use case.

> But personally, not having to escape < or > makes the code more readable for me.

But it is invalid XML (any XML editor should issue an error). It will eventually fail with the speedata Publisher. One day I will reactivate the other (Go based) XML parser which is more strict than my parser (but much safer). The one I use is a lousy piece of code.

(So, I am not trying to convince you to switch, but just give a warning that one day it will generate an error)

**Reply by pr-apes** [reply]
I didn't realize that `<` can be escaped with `not( > )`.

**Reply by pr-apes** [reply]
I mean, `<` is escaped with `not( >= )`.

**Reply by pgundlach** [reply]
> I mean, `<` is escaped with `not( >= )`.

No, you can't do this in XPath >= 2.0: `4 < (1,4,5) and 4 >= (1,4,5)` should give false but returns true.

**Reply by pr-apes** [reply]
Is this related to https://www.w3.org/XML/2007/qt-errata/xpath20-errata.html#E6 (https://www.w3.org/Bugs/Public/show_bug.cgi?id=5445)?

In that case, what seems to be problematic is comparing (other than equal and not equal) numbers and booleans.

As far as I know, I'm comparing number to number. And I'm afraid I don't understand how number and sequences may be compared.

Sorry if I'm missing your point, but I must admit this is all Greek to me.

**Reply by pgundlach** [reply]
I just wanted to state that except for simple cases `<` is not the equivalent of `not( ... >= .. )`.  Comparing numbers is ok though.

**Reply by pr-apes** [reply]
Many thanks for your reply, @pgundlach. Everything is clear now.

Here you have an improved version (that also fits letter paper [by shrinking it to the width):

``` xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
  xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Pageformat width="297mm" height="210mm"/>

  <Record element="data">
    <Loop select="if (sd:number-of-pages($pdfdoc) mod 4 != 0) then
        (sd:number-of-pages($pdfdoc) + (4 - (sd:number-of-pages($pdfdoc) mod 4))) div 2
      else sd:number-of-pages($pdfdoc) div 2"
    variable="runpage">
      <Switch>
        <Case test="sd:odd($runpage)">
          <Switch>
            <Case test="if (not((sd:number-of-pages($pdfdoc) + ((4-(sd:number-of-pages($pdfdoc)) mod 4) mod 4) - ($runpage - 1)) > sd:number-of-pages($pdfdoc))) then true() else">
              <PlaceObject column="{$_pagewidth div 4}" row="{$_pageheight div 2}" hreference="center" vreference="middle">
                <Image file="{$pdfdoc}" width="{$_pagewidth div 2}" page="{sd:number-of-pages($pdfdoc) + ((4-(sd:number-of-pages($pdfdoc)) mod 4) mod 4) - ($runpage - 1)}"/>
              </PlaceObject>
            </Case>
          </Switch>
          <PlaceObject column="{$_pagewidth * 0.75}" row="{$_pageheight div 2}" hreference="center" vreference="middle">
            <Image file="{$pdfdoc}" width="{$_pagewidth div 2}" page="{$runpage}"/>
          </PlaceObject>
          <ClearPage/>
          <Switch>
            <Case test="$runpage + 1 <= sd:number-of-pages($pdfdoc)">
              <PlaceObject column="{$_pagewidth div 4}" row="{$_pageheight div 2}" hreference="center" vreference="middle">
                <Image file="{$pdfdoc}" width="{$_pagewidth div 2}" page="{$runpage+1}"/>
              </PlaceObject>
            </Case>
          </Switch>
          <Switch>
            <Case test="if (not((sd:number-of-pages($pdfdoc) + ((4-(sd:number-of-pages($pdfdoc)) mod 4) mod 4) - $runpage) > sd:number-of-pages($pdfdoc))) then true() else">
              <PlaceObject column="{$_pagewidth * 0.75}" row="{$_pageheight div 2}" hreference="center" vreference="middle">
                <Image file="{$pdfdoc}" width="{$_pagewidth div 2}" page="{sd:number-of-pages($pdfdoc) + ((4-(sd:number-of-pages($pdfdoc)) mod 4) mod 4) - $runpage}"/>
              </PlaceObject>
            </Case>
          </Switch>
          <ClearPage/>
        </Case>
      </Switch>
    </Loop>
  </Record>

</Layout>
```

Document variable now reads `pdfdoc` and there were some wrong-nested parentheses that have been fixed.

**Comment by pr-apes** [comment]
>With a filter, it would be easy to create a table with the right page order, but I don't know how to pass the table to the layout and how to loop the page order.
>
> I have just discovered that `<ForAll>` is my friend here (after converting the table to a data source.

I took too much for granted, since there is no way to get the total number of pages from a PDF in a filter.

**Comment by cardonja** [comment]
I came across Speedata Publisher and this thread searching for a way to print A4 booklets with a bleedless spine cover.
Building on the concepts and code in this thread (and of course on the hard work of @pgundlach - Speedata Publisher for me is by far the best tool for repeatable and consistent results in database publishing), this is what I use for the imposition of A4-pages on A3 spreads with an A3-cover. Creep is hardcoded but can easily be adjusted, depending on paper thickness.
The result is ready for printing to any A3-MFP with saddle stitch finisher.
Just make sure to append three empty pages to your content.pdf with ClearPage.


```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
  xmlns:sd="urn:speedata:2009/publisher/functions/en">
  <SetGrid height="1pt" width="1pt"/>
  <Pagetype name="full" width="420mm" height="297mm" test="true()">
    <Margin left="0mm" right="0mm" top="0mm" bottom="0mm"/>
  </Pagetype>
  <SetVariable variable="cover" select="'cover.pdf'" />
  <SetVariable variable="pdfdoc" select="'content.pdf'"/>
  <SetVariable variable="numpages" select="floor(sd:number-of-pages($pdfdoc) div 4) * 2"/>
  <Record element="data">
    <PlaceObject row="0pt" column="0pt">
      <Image file="{$cover}" width="420mm"/>
    </PlaceObject>
    <ClearPage openon="right"/>
    <Loop select="$numpages" variable="runpage">
     <SetVariable variable="shift" select="floor($runpage div 3)"/>
     <SetVariable variable="leftpage" select="sd:alternating('left', {($numpages*2)-$runpage+1}, {$runpage})"/>
     <SetVariable variable="rightpage" select="sd:alternating('right', {$runpage}, {($numpages*2)-$runpage+1})"/>
     <PlaceObject column="{$shift}pt" row="0pt">
       <Image file="{$pdfdoc}" width="210mm" height="297mm" clip="yes" page="{$leftpage}"/>
     </PlaceObject>
     <PlaceObject column="{(sd:tounit('pt','210mm')-$shift)}pt" row="0pt">
       <Image file="{$pdfdoc}" width="210mm" height="297mm" clip="yes" page="{$rightpage}"/>
     </PlaceObject>
     <ClearPage/>
    </Loop>
  </Record>
</Layout>
```
